sdks/python/apache_beam/runners/common.py

```diff
@@ -771,22 +810,33 @@ def try_split(self, fraction):
       # Temporary workaround for [BEAM-7473]: get current_watermark before
       # split, in case watermark gets advanced before getting split results.
       # In worst case, current_watermark is always stale, which is ok.
-      if self.watermark_estimator:
-        current_watermark = self.watermark_estimator.current_watermark()
-      else:
-        current_watermark = None
+      current_watermark = None
       split = restriction_tracker.try_split(fraction)
       if split:
         primary, residual = split
         element = self.current_windowed_value.value
         restriction_provider = self.signature.get_restriction_provider()
         primary_size = restriction_provider.restriction_size(element, primary)
         residual_size = restriction_provider.restriction_size(element, residual)
-        return (
-            ((self.current_windowed_value.with_value((
-                (element, primary), primary_size)), None), None),
-            ((self.current_windowed_value.with_value((
-                (element, residual), residual_size)), current_watermark), None))
+        if self.threadsafe_watermark_estimator:
+          current_watermark = (
+              self.threadsafe_watermark_estimator.current_watermark())
+          estimator_state = (
+              self.threadsafe_watermark_estimator.current_estimator_state())
+          return (
+              ((self.current_windowed_value.with_value((
+                  (element, (primary, None)), primary_size)), None), None),
```

Comment:
Conceptually, yes. I'm thinking about whether we really need to keep the original estimator_state for primary. For now we only use estimator_state to re-create watermark_estimator, however, we should use original input to re-create watermark_estimator when this bundle is got retried.

Reply:
I think it's fine to re-create it rather than store it--no use caching away the original estimator state right after we create it just to put it here.